tests/functional/lifecycle/LifecycleTask.js

```diff
@@ -0,0 +1,770 @@
+const assert = require('assert');
+const uuid = require('uuid/v4');
+const util = require('util');
+
+const { errors } = require('arsenal');
+const Logger = require('werelogs').Logger;
+
+const LifecycleTask = require('../../../extensions/lifecycle' +
+    '/tasks/LifecycleTask');
+const Rule = require('../../utils/Rule');
+
+/*
+    NOTE:
+    When using CURRENT, running the lifecycle tasks will internally use
+    `Date.now` to compare dates. This means that CURRENT date usage will most
+    likely be in the past.
+    To avoid flakiness, I will be setting the day of CURRENT back 2 day to
+    avoid any flakiness.
+    When comparing rules, if I set a rule that says "Days: 1", then any objects
+    using CURRENT as LastModified should pass. To avoid flakiness, set a rule
+    that says "Days: 3" to have any objects using CURRENT as LastModified to
+    not pass.
+*/
+// current date
+const CURRENT = new Date();
+CURRENT.setDate(CURRENT.getDate() - 2);
+// 5 days prior to currentDate
+const PAST = new Date(CURRENT);
+PAST.setDate(PAST.getDate() - 5);
+// 5 days after currentDate
+const FUTURE = new Date(CURRENT);
+FUTURE.setDate(FUTURE.getDate() + 5);
+
+const OWNER = 'testOwner';
+
+// NOTE: Since all S3 calls are only 'GET's, I decided to just go for mock
+//  S3 that defines all buckets and their conditions
+class S3Mock {
```

Comment:
As a general comment, mocking is most useful when testing corner cases or unusual cases (like error injection). When it comes to emulating a real S3 behavior, it can be a better test to use the actual S3 process through functional tests, since creating mocks involves the risk of testing components which do not reflect the behavior of the real ones, as well as increasing complexity of writing and maintaining the mock code itself.

Reply:
I think as I was writing this, I realized how bulky it was getting and I also was thinking it'd be hard to maintain and extend :/

Will look into using S3 👍